### PR TITLE
Add a `maximumWidthRatio` property to customize maximum width of scale bar.

### PR DESCRIPTION
**Pull request by m-stephen**
This value is limited from 0.1 to 1.
Default value is 0.5

### CODE REVIEW

File: platform/ios/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ Mapbox welcomes participation and contributions from everyone. Please read [CONT
 ## master
 
 * Fix ornaments' view constraints bug when devices change to bold-text mode. ([#57](https://github.com/mapbox/mapbox-gl-native-ios/pull/57))
+* Add a `maximumWidthRatio` property for customizing maximum width of scale bar. ([#65](https://github.com/mapbox/mapbox-gl-native-ios/pull/65))
```
Comment (fabian-guerra) [inline platform/ios/CHANGELOG.md]: What do you think if we rephrase this to:
> Added `MGLMapView.maximumWidthRatio` property for customizing the width ratio of scale bar. ([#65](https://github.com/mapbox/mapbox-gl-native-ios/pull/65))

File: platform/ios/src/MGLMapView.h [hunk 2]
```diff
@@ -329,6 +329,13 @@ MGL_EXPORT
  */
 @property (nonatomic, assign) CGPoint scaleBarMargins;
 
+/**
+ The maximum width ratio of the scale bar relative to the map view's frame.
+ This value is limited from 0.1 to 1.
+ Default is 0.5(half of map view's width).
+ */
+@property (nonatomic, assign) CGFloat scaleBarMaximumWidthRatio;
```
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.h]: What do you think if we rename this property to `scaleBarWidthRatio`?

File: platform/ios/src/MGLMapView.h [hunk 3]
```diff
@@ -329,6 +329,13 @@ MGL_EXPORT
  */
 @property (nonatomic, assign) CGPoint scaleBarMargins;
 
+/**
+ The maximum width ratio of the scale bar relative to the map view's frame.
```
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.h]: Seems that is not actually a maximum because you don't have an explicit range. e.g. minimum = 0.1 maximum = 1.0
The behavior for this property is to specify the width ratio.

File: platform/ios/src/MGLMapView.h (hunk 2 above)
Comment (m-stephen) [reply platform/ios/src/MGLMapView.h]: I think if changing this name, it might lead a wrong meaning. 
The value will only affect the maximum width of scale bar but could not equal to the actual width. 
The actual depends on the zoom level, and we can only say the scale bar's actual width is always less than or equal to the maximum width. 
@fabian-guerra

File: platform/ios/src/MGLMapView.h (hunk 3 above)
Comment (m-stephen) [reply platform/ios/src/MGLMapView.h]: As above I mentioned, this value is not for the actual width ratio, but only limiting the maximum width of scale bar(the actual width is dynamic and depends on zoom level)

File: platform/ios/src/MGLMapView.mm [hunk 4]
```diff
@@ -800,9 +801,20 @@ - (void)setScaleBarPosition:(MGLOrnamentPosition)scaleBarPosition {
 - (void)setScaleBarMargins:(CGPoint)scaleBarMargins {
     MGLLogDebug(@"Setting scaleBarMargins: (x:%f, y:%f)", scaleBarMargins.x, scaleBarMargins.y);
     _scaleBarMargins = scaleBarMargins;
+    if([self.scaleBar isKindOfClass:[MGLScaleBar class]]) {
+        ((MGLScaleBar *)self.scaleBar).margins = scaleBarMargins;
+    }
     [self installScaleBarConstraints];
 }
 
+- (void)setScaleBarMaximumWidthRatio:(CGFloat)scaleBarMaximumWidthRatio {
```
Comment (tsuz) [inline platform/ios/src/MGLMapView.mm]: Can this test be included within `testChangeScaleBarSize`?

**Comment by julianrex** [comment]
I'm not sure this PR solves the problem in https://github.com/mapbox/mapbox-gl-native-ios/issues/52 

e.g. 

- Set a width ratio of 0.3. 
- Run on iPad.
- In portrait mode, scale bar is acceptable
- In landscape, developer may consider it still to be too wide.

Does it?

What do you think about adding an optional *absolute* maximum width, measured in points?

Or are you suggesting that the developer modifies the ratio on rotation?

File: platform/ios/src/MGLMapView.mm (hunk 4 above)
Comment (tsuz) [reply platform/ios/src/MGLMapView.mm]: Correct me if I'm wrong but I'm not sure if the tests cover the case where after this method is called, the scale bar adjusts accordingly.

**Comment by m-stephen** [comment]
> Does it?
> 
> What do you think about adding an optional absolute maximum width, measured in points?
> 
> Or are you suggesting that the developer modifies the ratio on rotation?

Since we cannot know developers' UI requirements, I think developer should modify the ratio manually. @julianrex 

If developers display a map on a iPhone and support orientation, but one developer want a full screen width scale bar, but another want a half one. They both looks good in portrait mode, but when in landscape mode, if we make a limit absolute maximum width, it might be too short/long and out of expectations.

File: platform/ios/CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 
 Mapbox welcomes participation and contributions from everyone. Please read [CONTRIBUTING.md](../../CONTRIBUTING.md) to get started.
 
+## Master
+
+* Add a `maximumWidthRatio` property for customizing maximum width of scale bar. ([#65](https://github.com/mapbox/mapbox-gl-native-ios/pull/65))
```
Comment (fabian-guerra) [inline platform/ios/CHANGELOG.md]: The blurb in changelogs should start with a verb in past tense. When you add a property you have to include the class name in order to link the documentation.
Please change the above to:
> Added \`MGLMapView.maximumWidthRatio\` property for customizing the width ratio of map view's scale bar.

File: platform/ios/src/MGLMapView.mm [hunk 6]
```diff
@@ -800,9 +801,21 @@ - (void)setScaleBarPosition:(MGLOrnamentPosition)scaleBarPosition {
...
```
Comment (fabian-guerra) [inline platform/ios/src/MGLMapView.mm]: Please add the call to `MGLLogDebug`

**Comment by chloekraw** [comment]
@m-stephen - is this ready for re-review?

**Review by habibahabiba77 — APPROVED** [review]

**Review by habibahabiba77 — COMMENTED** [review]
Please let me know when I can get my back account and how I do it my maps open location

**Review by habibahabiba77 — APPROVED** [review]
[](https://api.mapbox.com/styles/v1/habibahabiba77/ckh2y671y258019p9p7u6otli.html?fresh=true&title=copy&access_token=pk.eyJ1IjoiaGFiaWJhaGFiaWJhNzciLCJhIjoiY2tnM3psd29lMGZycDMxcDRxMWZyY2x0dyJ9.18NwOEay84kyXrEkDZX9Jg)

**Comment by lloydsheng** [comment]
@knov  @m-stephen has left. Does this function need to continue? I can take over the job.

/cc @julianrex

**Comment by tsuz** [comment]
@m-stephen  @lloydsheng Long over due. Does the scalebar adjust accordingly when it's in a SplitView?

**Comment by lloydsheng** [comment]
> Does the scalebar adjust accordingly when it's in a SplitView?

@tsuz  Per https://github.com/mapbox/mapbox-gl-native-ios/issues/52#issuecomment-565930806 and code, It should self-adjust.

**Review by tsuz — APPROVED** [review]
@lloydsheng  👍  Thank you!
